### ISSUE TITLE
False positive CA1508 with switch expression on nullable object

### ISSUE THREAD

**Issue by blouflashdb** [dotnet/roslyn-analyzers#7767]
**Area:** Code Analysis (CA1508)

**Version:**
*   .NET SDK Version: 9.0.306

**Description:**

When using a switch expression on a nullable object, the C# compiler's static analysis incorrectly flags a potential null access as "dead code" with a `CA1508` warning. When the code is modified to what the analyzer suggests, it then correctly raises a `CS8602` warning ("Dereference of a possibly null reference"). This indicates that the initial `CA1508` warning is a false positive.

**Steps to Reproduce:**

1.  Create a new C# project with nullable enabled.
2.  Add the following code:

```csharp
private static string? GetPrettyValue(object? value) =>
    value switch
    {
        DateTimeOffset dateTimeOffset => dateTimeOffset.ToString(
            "o"
        ),
        _ => value?.ToString(),
    };
```

**Expected Behavior:**

The code should compile without any warnings. The `value?.ToString()` is necessary because `value` can be `null` in the `_` arm of the switch expression.

**Actual Behavior:**

The code produces the following warning:
`CA1508: 'value' is never 'null'. Remove or refactor the condition(s) to avoid dead code.`

If the code is changed as suggested by the analyzer, by removing the conditional access operator (`?.`):

```csharp
private static string? GetPrettyValue(object? value) =>
    value switch
    {
        DateTimeOffset dateTimeOffset => dateTimeOffset.ToString(
            "o"
        ),
        _ => value.ToString(),
    };
```

A new, and in this case correct, warning is generated:
`CS8602: Dereference of a possibly null reference.`

This demonstrates that the `CA1508` warning in the first code example is a false positive. The analyzer is incorrectly concluding that `value` cannot be `null` in the `_` arm of the switch expression.
